### PR TITLE
recovery: add machinery for saving and rolling back congestion control state

### PR DESCRIPTION
**Pull request by ghedo**
This simply splits out the changes to save and restore CC state from https://github.com/cloudflare/quiche/pull/890 so e.g. it could also be used for https://github.com/cloudflare/quiche/pull/470.

### CODE REVIEW

File: src/recovery/cubic.rs [hunk 1]
```diff
@@ -646,3 +671,22 @@ mod tests {
         assert_eq!(r.cwnd(), cwnd_prev + r.max_datagram_size);
     }
 }
+
+fn checkpoint(r: &mut Recovery) {
```
Comment (junhochoi) [inline src/recovery/cubic.rs:675]: I think we just need to have a stub here and let each PR has actual implementation. Same to rollback()

File: src/recovery/cubic.rs [hunk 2]
```diff
@@ -75,6 +77,29 @@ pub struct State {
...
+
+    w_max: f64,
+
+    w_last_max: f64,
+
+    k: f64,
+
+    epoch_start: Option<Instant>,
+
+    lost_count: usize,
```
Comment (junhochoi) [inline src/recovery/cubic.rs:102]: This part can be removed too.

File: src/recovery/cubic.rs (hunk 1 above)
Comment (ghedo) [reply src/recovery/cubic.rs:675]: Why? This is the generic part from draft-ietf-tcpm-rfc8312bis, why would each PR re-implement the same thing?

**Review by junhochoi — APPROVED** [review]

**Comment by ghedo** [comment]
This was merged as part of https://github.com/cloudflare/quiche/pull/890, closing.

**Pull request closed without merging**